maven-core/src/main/java/org/apache/maven/DefaultMaven.java

```diff
@@ -463,51 +482,59 @@ private void validateLocalRepository( MavenExecutionRequest request )
         }
     }
 
-    private Collection<AbstractMavenLifecycleParticipant> getLifecycleParticipants( Collection<MavenProject> projects )
+    private <T> Collection<T> getExtensionComponents( Collection<MavenProject> projects, Class<T> role )
     {
-        Collection<AbstractMavenLifecycleParticipant> lifecycleListeners = new LinkedHashSet<>();
+        Collection<T> foundComponents = new LinkedHashSet<>();
 
         ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
         try
         {
             try
             {
-                lifecycleListeners.addAll( container.lookupList( AbstractMavenLifecycleParticipant.class ) );
+                foundComponents.addAll( container.lookupList( role ) );
             }
             catch ( ComponentLookupException e )
             {
                 // this is just silly, lookupList should return an empty list!
-                logger.warn( "Failed to lookup lifecycle participants: " + e.getMessage() );
+                logger.warn( "Failed to lookup " + role + ": " + e.getMessage() );
             }
 
-            Collection<ClassLoader> scannedRealms = new HashSet<>();
+            foundComponents.addAll( getProjectScopedExtensionComponents( projects, role ) );
+        }
+        finally
+        {
+            Thread.currentThread().setContextClassLoader( originalClassLoader );
+        }
+
+        return foundComponents;
+    }
+
+    protected <T> Collection<T> getProjectScopedExtensionComponents( Collection<MavenProject> projects, Class<T> role )
+    {
+
+        Collection<T> foundComponents = new LinkedHashSet<>();
+        Collection<ClassLoader> scannedRealms = new HashSet<>();
+
+        for ( MavenProject project : projects )
+        {
+            ClassLoader projectRealm = project.getClassRealm();
 
-            for ( MavenProject project : projects )
+            if ( projectRealm != null && scannedRealms.add( projectRealm ) )
             {
-                ClassLoader projectRealm = project.getClassRealm();
+                Thread.currentThread().setContextClassLoader( projectRealm );
```

Comment:
I'm far from familiar with how class loading works in Maven, but if we _change_ something here, shouldn't we _restore_ it by the end of the method (just like we do in `getExtensionComponents()`)?

Reply:
Good point, I'll check that.